Add scalafmt

mike-mcgann opened the pull request:
Adds dependencies for scalafmt/sbt-scalafmt, a format configuration, and a CI check to ensure source files are properly formatted.

[DAFFODIL-2133](https://issues.apache.org/jira/browse/DAFFODIL-2133)

stevedlawrence commented:
It might be useful to commit the actual scalafmt changes and we can see what we do and don't like. That would allow inline conversations if there's anything we aren't sure about or if we have different opinions.

mike-mcgann commented:
Formatting is now in the PR

stevedlawrence commented on daffodil-cli/src/main/scala/org/apache/daffodil/InfosetTypes.scala:
```diff
@@ -108,22 +108,22 @@ object InfosetType extends Enumeration {
-      case InfosetType.W3CDOM => W3CDOMInfosetHandler(dataProcessor)
-      case InfosetType.XML => XMLTextInfosetHandler(dataProcessor)
+      case InfosetType.W3CDOM    => W3CDOMInfosetHandler(dataProcessor)
+      case InfosetType.XML       => XMLTextInfosetHandler(dataProcessor)
```
Are there an opinions on this kind of alignment? I tend to not like becaause if we add a new longer case we have to reindent the whole thing and it makes the diff harder to see what actually changed.

mike-mcgann replied:
I personally like it. gofmt does the same with struct fields and makes everything look nice. You can omit whitespace changes in a github diff by clicking on the gear icon drop down, selecting "hide whitespace", and then clicking on apply.

stevedlawrence replied:
Yeah, I guess github does show clearly what is whitespace. Seems reasonable to me :+1:

mbeckerle replied:
I profoundly hate this aligning the arrows crap. 
Only looks good on toy situations.

E.g., something more real like:

```
foo match {
case S(bar, baz) if (f(baz) == g(bar)) => // ok
case T(bar)                                           => bar 
case _                                                   => Assert.invariantFailed("now I'm completely out of line space to put this message.")
}
```
As soon as some condition is slightly more complex than the others, the arrows all end up way off to the right and it is a mess.

mike-mcgann replied:
I've disabled all alignments with `align.preset = none`

stevedlawrence commented on daffodil-cli/src/main/scala/org/apache/daffodil/InfosetTypes.scala:
```diff
@@ -358,7 +356,7 @@ case class JDOMInfosetHandler(dataProcessor: DataProcessor)
 
 class JDOMInfosetParseResult(parseResult: ParseResult, output: JDOMInfosetOutputter)
-  extends InfosetParseResult(parseResult) {
+    extends InfosetParseResult(parseResult) {
```
This is interesting indenting extends 4 spaces instead of 2. Maybe that's a scala standard but I don't think we've been doing that. Not sure what the preference here.

mike-mcgann replied:
The default configuration tends to like 4 spaces on continuations like this. I changed a few of them from 4 to 2 but I may have missed a few.

mike-mcgann replied:
I'm not sure what the standard is, but I did find the parameter: `indent.extendSite`. I set it to 2 for now so you can see how that looks.

stevedlawrence reviewed (commented):
Skimmed through it and it seems reasonable. Only two things I noticed that are substantially different than we we already. I don't feel strongly about either style though.

Another though, can any IDE's import and use the .scalafmt .conf file or will they need to be configured to match? Or do we just make sure people run `sbt scalafmt` as part of the workflow?

mbeckerle commented on daffodil-core/src/main/scala/org/apache/daffodil/dpath/DFDLExpressionParser.scala:
```diff
@@ -270,34 +276,41 @@ class DFDLPathExpressionParser[T <: AnyRef](
 
-  def Predicate: Parser[PredicateExpression] = log(
-    "[" ~> Expr <~ "]" ^^ { PredicateExpression(_) })("predicate")
+  def Predicate: Parser[PredicateExpression] = log("[" ~> Expr <~ "]" ^^ { PredicateExpression(_) })("predicate")
```
The line length being used here is much too long.  (113 chars).

I think the max should be 96 chars / line.

mike-mcgann replied:
> Lines are far too long.

The default was 80 and it was reformatting a lot because there are quite a few long lines in the code. I bumped it up to 120 but that is quite long. I've moved it down to 96. See how that looks. 

> Lining up the arrows in match-case clauses is simply annoying and should be turned off.

Alignment has been disabled. 

> Changes are also needed to the way extends and with (one line per) for traits/classes are formatted.

What specifically would you like changed?

mbeckerle replied:
I withdraw the request about extends and with.

I took a look at the ElementBase trait and the indentation was not changed:

```
trait ElementBase
  extends Term
  with ElementLikeMixin
  with LocalElementMixin
  with Element_AnnotationMixin
  with NillableMixin
  with ElementBaseGrammarMixin
  with ElementRuntimeValuedPropertiesMixin
  with StringTextMixin
  with NumberTextMixin
  with CalendarTextMixin
  with BooleanTextMixin
  with TextNumberFormatMixin
  with EmptyElementParsePolicyMixin
  with TextStandardBaseMixin
  with OverlapCheckMixin
  with ElementBaseView {
```

So I'm fine with that.

mbeckerle reviewed (changes requested):
As much as I understand wanting to go with "the standard" way of formatting, I think this is unacceptable. "The standard" seems to be designed with getting little toy programs to look nice. 

Lines are far too long.

The indentation and new line break regimine seem to have more to do with how much can they get onto a line than what should be logically laid out differently. 

But not always. They pile all the extends and with statements of class trait definitions onto huge long lines, but they do seem to spread arguments of methods out one line per arg.  

Changes are needed to not make the lines so long are important.

Changes are also needed  to the way extends and with (one line per) for traits/classes are formatted. 

Lining up the arrows in match-case clauses is simply annoying and should be turned off.

tuxji commented:
I have tried to find a canonical "standard Scala code style" scalafmt configuration without any success.  It looks like various projects have put together their own differing scalafmt configurations, often overriding many tens of parameters.  I don't even know how well scalafmt's default configuration matches "standard Scala code style", that is, I don't know how many scalafmt parameters we will need to override to actually match modern standard Scala code style.  Oh well....

[Configuration | Scalafmt](https://scalameta.org/scalafmt/docs/configuration.html) documents the parameters (not very well, in my opinion) and shows their default values at the end of the page.

[Installation | Scalafmt](https://scalameta.org/scalafmt/docs/installation.html) says that Intellij IDEA will detect a .scalafmt.conf file at the root of the project, prompt you, and if you agree, will use scalafmt instead of Intellij's own formatter when formatting files.

[Gotchas | Scalafmt](https://scalameta.org/scalafmt/docs/gotchas.html) shows two situations where you can use newlines to avoid long lines without scalafmt overriding you even if maxColumn would force a longer line.

[Declarations | Style Guide](https://docs.scala-lang.org/style/declarations.html) shows what I think is the most important change  scalafmt should make to match the standard Scala code style (put constructors all on one line unless too long, then put each arg on its own line with trailing commas).  I want to see trailing commas used everywhere args are put on their own lines since trailing commas on args makes formatting look consistent and makes changes to the args much easier.

[DataBricks Scala Guide](https://github.com/databricks/scala-style-guide) has some different opinions than the standard Scala guide so I wouldn't follow everything it says, but some of what it says is worth reading.

[[incubator-pekko-grpc] 01/09: Update scalafmt](https://www.mail-archive.com/commits@pekko.apache.org/msg00325.html) shows the first message of a 9-part pull request formatting a Scala project.  It can serve as an example of how another project has done it.

I agree with Mike that alignment shouldn't be turned on.  I would like to see pragmatic changes to Daffodil code, that is, changes should be configured to be minimal except for one-line args; if scalafmt can change the last one-line arg to have a trailing comma, I'd prefer that to be done everywhere.  Thoughts?

mike-mcgann commented:
> It looks like various projects have put together their own differing scalafmt configurations, often overriding many tens of parameters. 

If there is a Scala project out there where we like the formatting, we can borrow that configuration. 

>  would like to see pragmatic changes to Daffodil code, that is, changes should be configured to be minimal except for one-line args

I've also enabled import rewriting to sort and expand imports to multiple lines. If these current changes are still too much, I can look into making it much more conservative. 

> if scalafmt can change the last one-line arg to have a trailing comma, I'd prefer that to be done everywhere. Thoughts?

I've enabled `rewrite.trailingCommas.style = always`.

mike-mcgann commented:
Code has been reformatted with the suggestions so far. Please take a look again and review.

tuxji commented:
> I've also enabled import rewriting to sort and expand imports to multiple lines. If these current changes are still too much, I can look into making it much more conservative.

I like consistency, so I don't consider sorting imports too much.  In fact, the current scalafmt configuration seems to have left 45% of Daffodil's Scala files unchanged and I see only whitespace changes to a few imports in the first 50% of the diffs (GitHub hides the remaining 50% of the diffs from view).  If scalafmt is really sorting the imports (can you check?), it must be matching the order already being used, so that's a promising start.

However, I'll find one of these whitespace changes and comment on them there since I think Mike will want to preserve these whitespace characters inside { } imports (he put explicit instructions in an IDE setup page to ensure these whitespace characters would be inserted).

tuxji commented on daffodil-core/src/main/scala/org/apache/daffodil/dpath/DFDLExpressionParser.scala:
```diff
@@ -17,8 +17,8 @@
 
-import java.math.{ BigDecimal => JBigDecimal }
-import java.math.{ BigInteger => JBigInt }
+import java.math.{BigDecimal => JBigDecimal}
+import java.math.{BigInteger => JBigInt}
```
Our Daffodil wiki's [Intellij IDEA Setup](https://cwiki.apache.org/confluence/display/DAFFODIL/IntelliJ+IDEA+Setup) has the following comment by Mike:

> Some of IDEA's Scala settings are not consistent with our formatting standard. On the Spaces tab, check boxes for "Import braces".

Therefore, is there a way to stop scalafmt from removing these whitespace characters from import braces?

mike-mcgann replied:
Yes, that would be `spaces.inImportCurlyBraces = true` and I have added that to the config.

tuxji commented on daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:
```diff
@@ -147,7 +146,10 @@ class CLIConf(arguments: Array[String]) extends scallop.ScallopConf(arguments) {
           }
-          case _ => Left("you should provide no more than one argument for this option") // Error because we expect there to be at most one flag
+          case _ =>
+            Left(
+              "you should provide no more than one argument for this option",
+            ) // Error because we expect there to be at most one flag
```
> I've enabled rewrite.trailingCommas.style = always.

I suspect that scalafmt applies this parameter only to function call arguments, not method parameters, because this is one of the only few places I've seen a trailing comma added.  Oh well, if that's all scalafmt can do, that's okay.

This particular Left() call, BTW, was formatted oddly because there was a long comment after the right parenthesis.  This is one of the (I hope only a few) places where we're going to have to move such a comment to its own line to get the Left() back on one line, like this:

```scala
          case _ =>
            // Error because we expect there to be at most one flag
            Left("you should provide no more than one argument for this option")
```

mike-mcgann replied:
So, I think for this we need the following:

```
danglingParentheses.callSite = true
danglingParentheses.defnSite = true
rewrite.trailingCommas.style = always
```

I've put that in the config so we can look at that.

tuxji replied:
Yes, v3 looks much better now!  Others, please take a look and review too.

tuxji commented on daffodil-cli/src/main/scala/org/apache/daffodil/InfosetTypes.scala:
```diff
@@ -643,7 +653,12 @@ case class SAXInfosetHandler(dataProcessor: DataProcessor, forPerformance: Boole
     def replay(h: ContentHandler): Unit = h.startDocument()
   }
 
-  case class SaxEventStartElement(uri: String, localName: String, qName: String, atts: Attributes) extends SaxEvent {
+  case class SaxEventStartElement(
+    uri: String,
+    localName: String,
+    qName: String,
+    atts: Attributes)
+    extends SaxEvent {
     def replay(h: ContentHandler): Unit = h.startElement(uri, localName, qName, atts)
   }
```
> I've enabled rewrite.trailingCommas.style = always.

It looks like scalafmt doesn't adds a trailing comma to the last constructor/method parameter despite what `rewrite.trailingCommas.style=always` says.  That's a shame because now the extends clause and the def statement have nothing visually separating them anymore.  If scalafmt had added the trailing comma to the last parameter, the ) and the extend clause would be on the same line and there would be a clear visual separation:

```scala
  case class SaxEventStartElement(
    uri: String,
    localName: String,
    qName: String,
    atts: Attributes,
  ) extends SaxEvent {
    def replay(h: ContentHandler): Unit = h.startElement(uri, localName, qName, atts)
  }
```

tuxji commented on .scalafmt.conf:
```diff
@@ -0,0 +1,27 @@
+# limitations under the License.
+
+version = 3.7.1
+runner.dialect = scala213
```
Oh, one more comment.  Daffodil is still using Scala 2.12.  We haven't upgraded to Scala 2.13 yet since many changes would be needed.  We should use the scala212 dialect here too.

mike-mcgann replied:
Updated

mbeckerle commented:
I disagree about rewrite.trailingCommas.style.

This should be never in my book.  

A comma followed by a closing parethesis... to my eye that's a syntax error and I'm immediately thinking what got cut out by accident and where did it go.

tuxji commented:
> I disagree about rewrite.trailingCommas.style.

Well, I agree with the current Scala [guide](https://docs.scala-lang.org/style/declarations.html) where it advises using trailing commas in declarations.  [SIP-27 Trailing Commas](https://docs.scala-lang.org/sips/trailing-commas.html) shipped in Scala 2.12.2, so it may have been a more recent Scala change than when Steve and you wrote most of Daffodil.  I think the reasons cited in that SIP are pretty compelling, don't you?

JavaScript also [allows](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas) trailing commands in array literals, object literals, parameter definitions, function calls, named imports, named exports, and array and object destructuring.  In all these cases, the trailing comma is entirely optional and doesn't change the program's semantics in any way.  The trailing comma is particular useful when adding, removing, or reordering items in a list that spans multiple lines, because it reduces the number of lines that need to be changed, which helps with both editing and reviewing the diff.  C# also [allows](https://www.jetbrains.com/help/rider/Trailing_Commas_Style.html) you to have the trailing comma after the last item.

I also found another place in the standard Scala documentation recommending using trailing commas after [compiler options](https://docs.scala-lang.org/overviews/compiler-options/#use-compiler-options-with-sbt) in build.sbt.

tuxji commented:
> A comma followed by a closing parenthesis... to my eye that's a syntax error and I'm immediately thinking what got cut out by accident and where did it go.

However, that's also a good point.  The Scala guide itself says never to use trailing commas followed by a closing parenthesis on the same line for that same reason.  Scalafmt must always place the closing parenthesis on the next line indented closer to the left than the list items so that we can visually see that the list is closed and finished.

mbeckerle commented:
> > A comma followed by a closing parenthesis... to my eye that's a syntax error and I'm immediately thinking what got cut out by accident and where did it go.
> 
> However, that's also a good point. The Scala guide itself says never to use trailing commas followed by a closing parenthesis on the same line for that same reason. Scalafmt must always place the closing parenthesis on the next line indented closer to the left than the list items so that we can visually see that the list is closed and finished.

I withdraw my objection to this based on the added comma being only in multi-line forms, and never immediately preceding a parenthesis on the same line.

stevedlawrence commented:
This needs to be rebased to fix the conflicts now that OSGI is merged.

Also, we need to figure out how to handle the windows failures. Git on windows enables autocrlf so all the files end in CRLF instead of LF and scalafmt complains. We could just add a condition so the check only runs on linux. Or maybe a better options is to split it off into another job? We really don't need to run scalafmt on multiple OSes--running once on ubuntu should be fine. We may want to also make the OSGI check and Rat check also be separate jobs, or maybe combine those three checks into some generic job that does checks that aren't os/environment specific?

mike-mcgann commented:
> This needs to be rebased to fix the conflicts now that OSGI is merged.

Updated

> Also, we need to figure out how to handle the windows failures. Git on windows enables autocrlf so all the files end in CRLF instead of LF and scalafmt complains. We could just add a condition so the check only runs on linux. Or maybe a better options is to split it off into another job?

I changed it to only run on linux. I think another job for checks seems a little heavy. It will probably take longer to install and prepare the environment than to run the tests.

tuxji commented on daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/DFDLEscapeScheme.scala:
```diff
@@ -36,8 +36,11 @@ import org.apache.daffodil.lib.schema.annotation.props.PropertyLookupResult
 import org.apache.daffodil.lib.schema.annotation.props.NotFound
 import org.apache.daffodil.lib.schema.annotation.props.Found
```
If you expand the diff to look at all the imports, you can see scalafmt did not sort the imports.  I wonder if something is missing from scalafmt's configuration.  BTW, when you try to change the configuration to sort the imports, I would suggest basing your changes on the v4 commit so you can run scalafmt and see from "git diff" more quickly whether scalafmt sorted the imports or not without seeing all the v4 changes obscuring the imports.

mike-mcgann replied:
This was a little confusing. There is a separate option that also needs to be included:

    rewrite.rules = [Imports]

and that only sorted imports there were on the same line. If I enable group sorting, then it sorts "vertically". I've set this to create three separate groups:

    rewrite.imports.groups = [
        ["scala\\..*", "java\\..*", "javax\\..*"],
        ["org\\.apache\\.daffodil\\..*"],
    ]

Standard libraries first, then daffodil, and followed by everything else.

stevedlawrence reviewed (approved):
+1 these changes seem reasonable to, nothing stands out as obviously wrong

tuxji commented on .scalafmt.conf:
```diff
@@ -0,0 +1,33 @@
+maxColumn = 96
+docstrings.style = keep
+indent.defnSite = 2
+indent.extendSite = 2
```
I notice that `indent.caseSite` (not overridden here yet) has 4 as its default size.  That parameter applies only to a long case expression separated by | which spans multiple lines and decides how much to indent the additional lines past the "case" on the first line.  Interestingly enough, the parameter's example shows a formatted case expression with `indent.caseSite = 5` which aligns all the parts of the expression.  It seems like the default value should've been 2 rather than 4 for consistency, but I'm not sure what actual Daffodil code is using.

The easiest way to find out would be to do a new commit with only a new `indent.caseSite = 2` line in this file and see what new changes appear in Daffodil files as a result.  If current Daffodil code is already indented 4 spaces (we would have to check the same files from the main branch before any scalafmt formatting was done), then we'll reject the change and keep using scalafmt's default value.

tuxji reviewed (changes requested):
Like Steve, I'm also happy with the changes scalafmt is making (especially how few changes there are overall), although there are so many files that it's hard to see all the changes.  However, we might want to add a few more lines to .scalafmt.conf before calling this PR done.  I would make each change one parameter at a time in its own commit so we can review the changes it makes and accept or reject these changes.  The additional parameters we might want to try are `indent.caseSite = 2`, `rewrite.rules = [ <add more rules one at a time> ]`, and maybe `rewrite.scala3.convertToNewSyntax = true`.

mike-mcgann commented:
>he additional parameters we might want to try are `indent.caseSite = 2`, `rewrite.rules = [ <add more rules one at a time> ]`, and maybe `rewrite.scala3.convertToNewSyntax = true`.

All the parameters have been added a single commit at a time except for `rewrite.scala3` as that requires the dialect to be set to scala3.

tuxji commented:
In conclusion, I like Imports and AvoidInfix, but we can remove `indent.caseSite = 2` and the rest of the rewrite rules.

mike-mcgann commented:
> In conclusion, I like Imports and AvoidInfix, but we can remove `indent.caseSite = 2` and the rest of the rewrite rules.

Updated with these latest changes

mbeckerle reviewed (approved):
+1

mbeckerle reviewed (commented):
When I try to use the .scalafmt.conf settings here in Intellij IDEA, I get an error unable to parse conf 

![image](https://user-images.githubusercontent.com/1524099/218766045-480ce585-ea8e-4b1d-bedd-cc962acac036.png)

mbeckerle commented:
Adding the version = 3.7.1 to the scalafmt.conf file fixes the IntelliJ IDEA settings problem.

tuxji commented:
Code coverage decreased by -0.41% only because some uncovered lines became more numerous after splitting across maxColumns.  I looked at the other pull requests and no one else seems to have an approved up-to-date pull request waiting to be merged, so let's merge this PR now before conflicts happen again.
